### PR TITLE
codeintel: Reuse gzip reader and writer

### PR DESCRIPTION
**Pull request by efritz**
Reduce allocations in gob serializer by reusing gzip reader and writer instances via a sync.Pool.

### CODE REVIEW

File: enterprise/internal/codeintel/bundles/persistence/serialization/gob/serializer.go [hunk 1]
```diff
@@ -16,48 +17,56 @@ func init() {
...
+	writers *sync.Pool
+}
 
 var _ serialization.Serializer = &gobSerializer{}
 
 func New() serialization.Serializer {
-	return &gobSerializer{}
+	return &gobSerializer{
+		readers: &sync.Pool{New: func() interface{} { return new(gzip.Reader) }},
+		writers: &sync.Pool{New: func() interface{} { w, _ := gzip.NewWriterLevel(nil, gzip.BestCompression); return w }},
```
Comment (ryanslade) [inline enterprise/internal/codeintel/bundles/persistence/serialization/gob/serializer.go:30]: Do we need `BestCompression`? Maybe `DefaultCompression` is enough and will be faster?

File: enterprise/internal/codeintel/bundles/persistence/serialization/gob/serializer.go [hunk 2]
```diff
@@ -16,48 +17,56 @@ func init() {
...
-func (*gobSerializer) UnmarshalDocumentData(data []byte) (document types.DocumentData, err error) {
-	r, err := gzip.NewReader(bytes.NewReader(data))
-	if err != nil {
+func (s *gobSerializer) UnmarshalDocumentData(data []byte) (document types.DocumentData, err error) {
+	r := s.readers.Get().(*gzip.Reader)
+	defer s.readers.Put(r)
+
+	if err := r.Reset(bytes.NewReader(data)); err != nil {
 		return types.DocumentData{}, err
 	}
```
Comment (ryanslade) [inline enterprise/internal/codeintel/bundles/persistence/serialization/gob/serializer.go:71]: I think we should be calling `Close` on the readers after we've finished encoding

https://golang.org/pkg/compress/gzip/#Reader.Close

File: enterprise/internal/codeintel/bundles/persistence/serialization/gob/serializer.go [hunk 3]
```diff
@@ -88,18 +101,19 @@ func (*gobSerializer) UnmarshalLocations(data []byte) (locations []types.Locatio
 }
 
 // compress gzips the bytes in the given reader.
-func compress(r io.Reader) ([]byte, error) {
+func (s *gobSerializer) compress(r io.Reader) ([]byte, error) {
+	w := s.writers.Get().(*gzip.Writer)
+	defer s.writers.Put(w)
+
```
Comment (ryanslade) [inline enterprise/internal/codeintel/bundles/persistence/serialization/gob/serializer.go:107]: Maybe wan reuse the buffers too?

Comment (efritz) [reply enterprise/internal/codeintel/bundles/persistence/serialization/gob/serializer.go:72]: Good suggestion, but this didn't buy us anything in the benchmark. I'll be updating this in a future PR to pass in a buffer to use (which is more controlled).

**Comment by efritz** [comment]
@keegancsmith What encoder? One wasn't suggested (just that gob was bad).

**Comment by efritz** [comment]
I'll work on updating the encoder use (or replacing it) in another PR.

**Comment by keegancsmith** [comment]
Yeah save for another PR. The reddit commit linked to a benchmark and it seems they ended up on msgpack. Our use case is pretty damn simple to be honest, so there are lots of options. I'd preferablly like whatever we pick doesn't introduce onerous requirements on our dev environment. Also on that note, is it even worth investigating right now? I am guessing we have bigger fish to fry vs somewhat small gains here.

**Comment by efritz** [comment]
I’m only investigating a change in format if I’m blocked on reviews, so it’s not a focus. A quick experiment showed msgp is not any faster or smaller than GOB for us. Code isn’t much cleaner either (requires a generate step but is otherwise kind of identical).

A bigger win would be adding reusable byte buffers to the serializer, which requires a small change in the interface of a few things.

**Pull request merged**